sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -390,7 +390,12 @@ object ColumnPruning extends Rule[LogicalPlan] {
   /** Applies a projection only when the child is producing unnecessary attributes */
   private def prunedChild(c: LogicalPlan, allReferences: AttributeSet) =
     if ((c.outputSet -- allReferences.filter(c.outputSet.contains)).nonEmpty) {
-      Project(allReferences.filter(c.outputSet.contains).toSeq, c)
+      c match {
+        case BroadcastHint(p) =>
+          BroadcastHint(Project(allReferences.filter(c.outputSet.contains).toSeq, p))
```

Comment:
Is there any other operator will be inserted between Join and BroadcastHint?

Maybe we could have a rule to pull up BroadcastHint until Join.